packages/yew/src/virtual_dom/vlist.rs

```diff
@@ -35,29 +74,235 @@ impl VList {
+        }
         self.children.push(child);
     }
 
     /// Add multiple `VNode` children.
     pub fn add_children(&mut self, children: impl IntoIterator<Item = VNode>) {
-        self.children.extend(children);
+        let it = children.into_iter();
+        let bound = it.size_hint();
+        self.children.reserve(bound.1.unwrap_or(bound.0));
+        for ch in it {
+            self.add_child(ch);
+        }
+    }
+
+    /// Recheck, if the all the children have keys.
+    ///
+    /// Run this, after modifying the child list that contained only keyed children prior to the
+    /// mutable dereference.
+    pub fn recheck_fully_keyed(&mut self) {
+        self.fully_keyed = self.children.iter().all(|ch| ch.has_key());
+    }
+
+    /// Diff and patch unkeyed child lists
+    fn apply_unkeyed(
+        parent_scope: &AnyScope,
+        parent: &Element,
+        mut next_sibling: NodeRef,
+        lefts: &mut [VNode],
+        rights: Vec<VNode>,
+    ) -> NodeRef {
+        let mut diff = lefts.len() as isize - rights.len() as isize;
+        let mut lefts_it = lefts.iter_mut().rev();
+        let mut rights_it = rights.into_iter().rev();
+
+        macro_rules! apply {
+            ($l:expr, $r:expr) => {
+                test_log!("parent={:?}", parent.outer_html());
+                next_sibling = advance_next_sibling!(next_sibling, || $l.apply(
+                    parent_scope,
+                    parent,
+                    next_sibling,
+                    $r
+                ));
+            };
+        }
+
+        // Add missing nodes
+        while diff > 0 {
+            let l = lefts_it.next().unwrap();
+            test_log!("adding: {:?}", l);
+            apply!(l, None);
+            diff -= 1;
+        }
+        // Remove extra nodes
+        while diff < 0 {
+            let mut r = rights_it.next().unwrap();
+            test_log!("removing: {:?}", r);
+            r.detach(parent);
+            diff += 1;
+        }
+
+        for (l, r) in lefts_it.zip(rights_it) {
+            test_log!("patching: {:?} -> {:?}", r, l);
+            apply!(l, r.into());
+        }
+
+        next_sibling
     }
 
-    fn children_keys(&self, warn: bool) -> HashSet<Key> {
-        let mut hash_set = HashSet::with_capacity(self.children.len());
-        for l in self.children.iter() {
-            if let Some(k) = l.key() {
-                if !hash_set.insert(k.clone()) && warn {
-                    log::warn!("Key '{}' is not unique in list but must be.", k);
+    /// Diff and patch fully keyed child lists.
+    ///
+    /// Optimized for node addition or removal from either end of the list and small changes in the
+    /// middle.
+    fn apply_keyed(
+        parent_scope: &AnyScope,
+        parent: &Element,
+        mut next_sibling: NodeRef,
+        lefts: &mut [VNode],
+        rights: Vec<VNode>,
+    ) -> NodeRef {
+        use std::mem::{transmute, MaybeUninit};
+
+        macro_rules! map_keys {
```

Comment:
I'm not sure about the benefit of using a macro here (especially with an eye towards compile times).

Reply:
It saves us from copy-pasting and I think the compile costs are negligible. Especially so, with how much the `html!` procedural macro does.